xarray/core/concat.py

```diff
@@ -590,7 +590,7 @@ def get_indexes(name):
             # preserves original variable order
             result_vars[name] = result_vars.pop(name)
 
-    result = Dataset(result_vars, attrs=result_attrs)
+    result = cast(T_Dataset, Dataset(result_vars, attrs=result_attrs))
```

Comment:
Here the actual result value is hardcoded to be of type `Dataset` which would not match `T_Dataset` if `datasets` contains some subclass of `xarray.Dataset`.

Suggestion to try (which may not need the cast): 
`result = type(datasets[0])(result_vars, attrs=result_attrs)`

Reply:
Nice find! I do wonder if we have this issue in more places?

Reply:
Yes this is quite endemic to hardcode Dataset / DataArray, rather than use the class of the argument. 

We would need to make some widespread changes for this to work correctly — I think that would be welcome. We'd probably need to find a good way to test this across methods too (though maybe mypy could help there?).

And we might need a property of `Dataset` for what to class to use when returning a `DataArray`, which becomes more complicated.

Reply:
Looks like this broke somebody's subclass that doesn't support the `attrs` argument: https://github.com/pydata/xarray/issues/6827